framework/source/class/qx/ui/form/VirtualSelectBox.js

```diff
@@ -445,6 +467,151 @@ qx.Class.define("qx.ui.form.VirtualSelectBox",
       var d = event.getData();
       var old = (d.removed.length ? d.removed[0] : null);
       this.fireDataEvent("changeValue", d.added[0], old);
+    },
+
+
+    /*
+    ---------------------------------------------------------------------------
+      INCREMENTAL SEARCH
+    ---------------------------------------------------------------------------
+    */
+
+    __filterValue : null,
+    __lastMatch : '',
+    __filterUpdateRunning : 0,
+    __filterInput : null,
+    __highlightStyle : null,
+    __lastRich : false,
+
+
+     __addFilterInput : function()
+    {
+      var input = this.__filterInput = new qx.ui.form.TextField().set({
+          appearance : 'widget',
+          liveUpdate : true,
+          height     : 0,
+          width      : 1 // must be > 0
+      });
+      this._add(input);
+      input.addListener("focus", function(e) {
+          // reopen after focus loss
+          this.open();
+      }, this);
+      this.addListener("focus", function(e) {
+          // move focus into TextField
+          input.focus();
+      }, this);
+      input.addListener('changeValue', function(e) {
+          if (this.__filterUpdateRunning === 0) this.__updateDelegate();
+      }, this);
+      input.addListener("keypress", function(e) {
+          switch (e.getKeyIdentifier()) {
+          case 'Enter':
+          case 'Escape':
+              this.__updateDelegate();
+              break;
+          // prevent cursor from being moved to beginning/end of input field
+          case 'Down':
+          case 'Up':
+              e.preventDefault();
+              break;
+          }
+      }, this);
+    },
+
+    __applyHighlightStyle : function(value, old) {
+      var styles = [];
+      if (value != null) {
+        var keys = Object.keys(value);
+        var k;
+        for (k=0; k< keys.length; k++) {
+          var key = keys[k].replace(/[A-Z]/g, m => "-" + m.toLowerCase());
+          styles.push(key + ':' + value[keys[k]]);
+        }
+        this.__highlightStyle = styles.join(';');
+      }
+      else {
+        this.__highlightStyle = '';
+      }
+    },
+
+    _highlightFilterValue : function (item)
+    {
+      return this.__highlightStyle ? item.replace(item, '<span style="' + this.__highlightStyle + '">' + qx.module.util.String.escapeHtml(this.__filterValue) + '</span>')
```

Comment:
This is good. It solves the problem... except in the case where the user had turned on `rich` intentionally, in which case we shouldn't escape it. So something like this, maybe:

```
  return (
    this.__highlightStyle
      ? item.replace(
          item,
          [
            '<span style="',
            this.__highlightStyle,
            '">',
            this.__lastRich
              ? qx.module.util.String.escapeHtml(this.__filterValue)
              : this.__filterValue,
            '</span>'
          ].join(""))
      : item);
```

Reply:
Done.

Reply:
Get highlight style via theming appearance and added the style to all the themes.

Reply:
hmm if there is 'evil' html outside the highlighted part, it will not get escaped I think

Reply:
> hmm if there is 'evil' html outside the highlighted part, it will not get escaped I think

Yes, as it wouldn't if the programmer has set them to rich himself (which is done exactly like here in one of the demobrowser examples) and doesn't take care what content he has in the model.